app.js

```diff
@@ -174,18 +183,49 @@ var jsonp = function(response) {
       var array = req.body;
 
       if (typeof(array) !== 'object') {
-        return res.send(400, util.format('Expect an object body, got[%s]', typeof(array)));
+        return res.send(400, util.format('Expected an object body, got[%s]', typeof(array)));
       }
 
-      if (! Array.isArray(array)) {
+      if (!Array.isArray(array)) {
         array = [array];
       }
 
       var count = 0;
       var duplicates = [];
+
       async.waterfall(
         [
-          lookupGroupId.bind(null, userid),
+          function(cb) {
+            // if no groupId was specified, just continue to upload for the
+            // connected user
+            if (!req.params.groupId) {
+              return cb(null, userid);
+            }
+
+            // get the groups for the logged-in user
+            gatekeeperClient.groupsForUser(userid, function(err, groups) {
+              if (err) {
+                return cb(err);
+              }
+
+              // and check them all to see if we have upload permissions
+              for (var id in groups) {
+                var group = groups[id];
+
+                if (id === req.params.groupId && (group.upload || group.root)) {
+                  return cb(null, id);
+                }
+              }
+
+              cb({
+                statusCode: 500,
+                message: 'You don\'t have rights to upload to that account.'
+              });
+            });
+          },
+          function(id, cb) {
```

Comment:
just checking but is this a groupId or userId? just so it reads consistently with the others